### PR TITLE
zy_hnerv

### PR DESCRIPTION
**Pull request by zaidyunis**
# submission name:
zy_hnerv

# upload zipped `archive.zip`
[archive.zip](https://github.com/user-attachments/files/31996663/archive.zip)

# report.txt
=== Evaluation config === batch_size: 16 device: cpu num_threads: 2 prefetch_queue_depth: 4 report: submissions/zy_hnerv/report.txt seed: 1234 submission_dir: submissions/zy_hnerv uncompressed_dir: /home/zaidyunis/comma_video_compression_challenge/videos video_names_file: /home/zaidyunis/comma_video_compression_challenge/public_test_video_names.txt === Evaluation results over 600 samples === Average PoseNet Distortion: 0.00132028 Average SegNet Distortion: 0.00794935 Submission file size: 277,090 bytes Original uncompressed size: 37,545,489 bytes Compression Rate: 0.00738011 Final score: 100segnet_dist + √(10posenet_dist) + 25*rate = 1.09
```

# does your submission require gpu for evaluation (inflation)?
No, inflate.py is CPU-only torch; thread count is tunable via INFLATE_THREADS

# did you include the compression script? and want it to be merged?
Yes to both

# is this submission competitive or innovative? explain why
Not gonna lie no, but I had fun doing this

# additional comments
As my first challenge for [comma.ai](http://comma.ai), I had a great experience grappling with the problem and building a workable solution.
My first step was to have a solid understanding of what the problem actually was. Essentially, [comma.ai](http://comma.ai) wants to compress videos to be as small as possible, while still preserving information that is essential to the workability of self-driving cars. In order to determine the effectiveness of a solution, a function is used and lower score equates to a more effective video.
The score is based on three factors: PoseNet, SegNet, and compression rate. Segnet takes the second frame of each two-frame pair and assigns each pixel to one of five semantic classes. It then compares the class map from the original frame with the reconstructed frame. Then the distortion is given by the number of mismatched semantic pixels / total pixels. PoseNet penalizes changes to motion over consecutive frames using mean-squared error. Compression rate essentially measures how small the result video is.
One thing that stuck out to me is that SegNet only looks at half the frames. My original thought would be that there could be a lot of value in finding out how to separate odd and even frames.
My original idea for a solution was a classical codec approach and was to adjust and test the coefficients in [compress.sh](http://compress.sh) and [inflate.py](http://inflate.py). I wanted to see if messing around with different factors would affect the performance of the model. I tested out the scale of the compression, the resolution, constant rate factor, bilinear vs bicubic, slow/ultrafast, and GOP(keyframes). However, upon testing different variables, one at a time, I realized that the score usually remained about the same or went up. I also realized that this wouldn't be an efficient way to truly reduce the score and I needed a faster way to test while finding a way to actually reduce the score.
This is when I looked into HNeRV, or Hybrid Neural Representation for Videos. Here, instead of simply encoding and decoding each frame, the video is treated as something that can be generated by a neural network. This type of encoder would be able to optimize the shared neural weights to predict the 96-dimensional latent code for each frame pair. This would work well because the same network weights can describe all 1,200 frames and the network would be able to learn recurring structures.
One drawback, however, is the fact that since the model is being trained on a single video, it is highly overfitted to this specific video. As a result, this same model probably wouldn't work amazingly on a driving video that has different features. Another thing that limited me was the environment I was training in, which only included 2 cores. To fully get the lowest score possible, it would need to run for a while on multiple GPUS.
Before writing and training code, I wanted to know what was actually possible. The approach that I settled on was to take the original frames, damage them in a specific way, and run them through the scoring function. This would tell me what design decisions would cost me, which I used going forward. Because my machine was slow, I looked into having the network generate smaller images and scale them up. This would have made training about three times cheaper, but the probe showed it would cost me 0.95 points, so I dropped the idea before spending hours on it. Additionally, I learned that most of the penalty was coming from PoseNet and not SegNet. This makes sense once you think about it, since motion between two frames lives in very small pixel shifts, and any blurring destroys it. SegNet, on the other hand, is more forgiving, since a smoother image could segment into roughly the same regions.
The final result was a score of 1.09, down from the baseline of 4.39. I know this is definitely not one of the top scores, but I am proud of the work that I was able to accomplish and I look forward to tackling more challenges in the future.

### CODE REVIEW

**Comment by YassineYousfi** [comment]
glad you had fun!
